### DISCUSSION TITLE
Failed to build DirectShow graph  for AXIS Media Viewer

### DISCUSSION THREAD

**Discussion by dholgaurav** [AxisCommunications/discussions#75] [Axis Windows SDKs] [answered]
Hi,

I am trying to run DewarpDirectShowCSharp sample application on our server. it is giving "Failed to build DirectShow graph" on the server.

Please note : I am able to run the same sample application on local machine, it is working as expected.

Please suggest how we can resolve this problem or what could be the missing

<img width="118" alt="error" src="https://user-images.githubusercontent.com/2485642/193631088-7ce40749-200e-4fc5-be37-e19b5a4b0222.png">

Regards,
Gaurav

**Comment by anders-axis** [comment]
Are you planing to do rendering also on server?

- You can try to change renderer? Easiest is to use the "AdvancedControl" sample and change to:
`viewer.VideoRenderer = (int)AMV_RENDER_OPTIONS.AMV_VIDEO_RENDERER_EVR;`

- You can also try to disable rendering by changing code to:
`viewer.Init(0, fileParser.MediaTypeBuffer, (long)IntPtr.Zero);`

**Comment by fre3de** [comment]
Desktop experience must be installed on Server OS or the renderers won't be available.

Go to Start >> Administrative Tools >> Server Manager.
On the left side pane, click on Features. ...
On the Select Features page, select the Desktop Experience check box.

You can try to playback a video clip in Windows media player, it uses the same renderer (EVR).

**Comment by dholgaurav** [comment]
Hi @anders-axis  and @fre3de 

I tried to enable Desktop Experience check box but then also it is not working.

After lost of experiment and I found the solution when I run the code with "VMR9 = true" it is not running on the server and when I make "VMR9= false" it is also working on server.

Can you please suggest what is the use of VMR9 parameters, into sample by default it is true

Regards,
Gaurav

**Comment by anders-axis** [comment, marked as answer]
Hi, first note that the VMR9 property is deprecated - you should use the VideoRenderer property if possible.

VMR9 = false, is the same as VideoRenderer = (int)AMV_RENDER_OPTIONS.AMV_VIDEO_RENDERER_VMR7.
The default renderer is VMR7 because of backwards-compatibility with older versions of the SDK.

I would recommend using EVR option in most cases, but it depends on different use-cases.

From the SDK documentation (VideoRenderer):

> This property does not effect video rendering when streaming MJPEG over HTTP. 
> 
> See Microsoft [DirectShow](https://github.com/orgs/AxisCommunications/discussions/DirectShow.html) documentation for more information about the Video Mixing Renderers and the Enhanced Video Renderer. 
> 
> The Enhanced Video Renderer are only available on Windows 7 and above. 
> 
> Some graphics adaptors have difficulties rendering very large image streams (> 3 MP) using VMR7. 
> 
> The Enhanced Video Renderer might improve pixel interpolation when video is scaled on Windows 7 and above. 
> 
> Video Mixing Renderer 9 is recommended for scrubbing.

**Reply by dholgaurav** [reply]
Thanks @anders-axis 

I took the reference from AXIS Dewarping SDK Samples (DewarpDirectShowCSharp) but I will  try to change the implementation and let you know if I face any problem
